[WIP] Planning_scene_monitor (current_state_monitor)

anasarrak opened the pull request.

anasarrak committed:
Prepare moveit_ros_planning to build for ROS2

anasarrak committed:
Update planning CMakelists to compile collision_plugin_loader

anasarrak committed:
Update collision_plugin_loader CMakelist

anasarrak committed:
Port collision_plugin_loader

anasarrak committed:
Update CMakeLists.txt

anasarrak committed:
Reverta back changes perception CMakelists

anasarrak committed:
Prepare cmakelists current_state_monitor

anasarrak committed:
Change the different calls
It's a still WIP, problems with tf2_buffer_ there is no _removeTransformsChangedListener and _addTransformsChangedListener
on geometry2 for ros2, think how to change it.

the appropriate change for ros::WallTime is std::chrono::system_clock since it's not implemented yet
find something to replace ros::WallDuration

anasarrak commented on moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp:
```diff
@@ -136,21 +138,26 @@ void planning_scene_monitor::CurrentStateMonitor::clearUpdateCallbacks()
     if (tf_buffer_ && !robot_model_->getMultiDOFJointModels().empty())
     {
       tf_connection_.reset(new TFConnection(
-          tf_buffer_->_addTransformsChangedListener(boost::bind(&CurrentStateMonitor::tfCallback, this))));
```
@davetcoleman and  @mlautman I'm trying to replace the _addTransformsChangedListener call since on ros2 geometry this function is not implemented, I would like some input about how to procede with that. 

There is a similar function called [addTransformableCallback](http://docs.ros.org/jade/api/tf2/html/classtf2_1_1BufferCore.html#a71a56dfdd102295b339684406218f986) but it has nothing to do with that.

anasarrak replied:
ping @henningkayser, @davetcoleman, @mlautman

davetcoleman replied:
Do you know why ``_addTransformsChangedListener()`` is not implemented?

@IanTheEngineer made this change 10 months ago, here: https://github.com/ros-planning/moveit/commit/736251dc77a24877135522eb901ae40a1ea9872b

Looking at that [PR](https://github.com/ros-planning/moveit/pull/830), it required the following PRs to pass:

This PR depends on the following upstream API changes for Melodic:
https://github.com/ros/geometry/pull/163 (Exposing `tf2_ros::Buffer` object from TransformListener)
https://github.com/ros/geometry2/pull/292 (tf2 Type conversions)
https://github.com/ros/geometry2/pull/294 (Additional tf2 Type conversions)
https://github.com/ros-visualization/rviz/pull/1236 (Exposing tf2 without ROS deprecation warnings)

So likely one of those needs to be forward ported to ROS2.

anasarrak replied:
> Do you know why `_addTransformsChangedListener()` is not implemented?

Look at the https://github.com/ros-planning/moveit/pull/830#issuecomment-389881823, on the porting procedure they just left it apart as it seems to be "deprecated".

anasarrak replied:
https://github.com/ros-planning/moveit/pull/830#issuecomment-386517416
> addTransformsChangedListener API has been replaced by the TransformableRequest interface. An example of using it is in the tf2::MessageFilters: http://docs.ros.org/jade/api/tf2_ros/html/c++/message__filter_8h_source.html

mlautman replied:
@IanTheEngineer can you help shed some light on this question?

anasarrak replied:
ping @mlautman, @davetcoleman 

I have to finish this...

mlautman replied:
@anasarrak I don't know if I have any more information than you do at this point.. Have you had the chance to hunt down the dependencies that Dave referenced to see if any of them might be holding up the port? Perhaps @tfoote's comment can help shed some light here: https://github.com/ros-planning/moveit/pull/830#issuecomment-386517416

anasarrak commented on moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp:
```diff
@@ -214,90 +226,99 @@ bool planning_scene_monitor::CurrentStateMonitor::haveCompleteState(std::vector<
 
-bool planning_scene_monitor::CurrentStateMonitor::waitForCurrentState(const ros::Time t, double wait_time) const
-{
-  ros::WallTime start = ros::WallTime::now();
```
@wjwwood As you mentioned on https://github.com/ros-planning/moveit2/issues/31 the walltime calls should be replaced with ``std::chrono::system_clock``, so I guess that the correct use of WallDuration should be something like ``std::chrono::duration`` or is there a better way to do it, like just using ``rclcpp::Duration`` calls?

wjwwood replied:
`std::chrono::duration` is sufficient I think. We were supposed to have to/from functions for converting between `std::chrono::duration` and `rclcpp::Duration`, but I don't know if they got implemented or not. It would be a good thing to contribute if they're missing.

I don't know what the context here is, so I can't advise you. Please give me an exact question or scenario and I can help more.

anasarrak replied:
Basically the question was: What is the best replacemanet for ros::WallDuration/ros::WallTime since they are not implemented for ros2. You have already answered me with, use _std::chrono::system_clock_ as replacement for _ros::walltime_ and _std::chrono::duration_ for _ros::WallDuration_, right?

anasarrak committed:
Update waitForCurrentState function adding std::chrono::duration

anasarrak committed:
change duration_cast to nanoseconds

anasarrak committed:
Use change boost::mutex to std::unique_lock<std::mutex>

anasarrak committed:
Replace boost::function to std::function
Much of the functionality that provides boost are implemented on C++11

davetcoleman commented on moveit_ros/planning/CMakeLists.txt:
```diff
@@ -47,52 +43,56 @@ set(THIS_PACKAGE_INCLUDE_DIRS
+ament_export_include_directories(include)
+ament_export_libraries(${libraries})
+
+ament_package()
```
A stray ``COLCON_IGNORE`` was committed in the following file:
``moveit_ros/planning/COLCON_IGNORE``

davetcoleman commented on moveit_ros/planning/planning_scene_monitor/CMakeLists.txt:
```diff
@@ -1,18 +1,29 @@
+  moveit_collision_plugin_loader
+  ${rclcpp_LIBRARIES})
+
+#TODO: (anasarrak) Fix this once moveit_ros_planning is finished
```
```suggestion
#TODO: (anasarrak) Fix this once moveit_ros_planning is converted to ROS2
```

If you forget to do this, let's have descriptive comments so others know what you meant by 'finished'

davetcoleman commented on moveit_ros/planning/planning_scene_monitor/include/moveit/planning_scene_monitor/current_state_monitor.h:
```diff
@@ -117,21 +120,23 @@ class CurrentStateMonitor
    *  @return Returns the current state */
   robot_state::RobotStatePtr getCurrentState() const;
 
+  std::pair<robot_state::RobotStatePtr, rclcpp::Time> getCurnode_rentStateAndTime() const;
```
the name of this function is broken. also, is this function needed?

anasarrak replied:
Fixed here https://github.com/AcutronicRobotics/moveit2/commit/f12ad0e007bd6d71004286659c56e9b43b374307, thanks for this nice catch on the mispelling (it's still a WIP). Also, it's needed on https://github.com/AcutronicRobotics/moveit2/blob/f12ad0e007bd6d71004286659c56e9b43b374307/moveit_ros/planning/planning_scene_monitor/src/trajectory_monitor.cpp#L110

davetcoleman commented on moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp:
```diff
@@ -163,36 +168,41 @@ void planning_scene_monitor::CurrentStateMonitor::stopStateMonitor()
   if (state_monitor_started_)
   {
-    joint_state_subscriber_.shutdown();
+    rclcpp::shutdown();
```
this is the wrong kind of shutdown. we don't want to turn off the entire node!

anasarrak replied:
Yes, I already had some changes for that, using ``joint_state_subscriber_.reset()``

davetcoleman committed:
Update moveit_ros/planning/planning_scene_monitor/CMakeLists.txt

Update TODO for the demos

Co-Authored-By: anasarrak <anasarrak@gmail.com>

anasarrak committed:
current_state_monitor.h miss spelling

anasarrak committed:
Fix current_state_monitor misspell

anasarrak committed:
current_state_monitor, update subscriber to shared

anasarrak committed:
Changes on publish/subscribers, logs and services

anasarrak committed:
planning_scene_monitor, change parameters on configureCollisionMatrix

anasarrak committed:
planning_scene_monitor Changes on parameters

anasarrak commented on moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp:
```diff
@@ -1349,39 +1349,60 @@ void PlanningSceneMonitor::configureCollisionMatrix(const planning_scene::Planni
-      RCLCPP_WARN(logger, "default_collision_operations is not an array");
-      return;
-    }
+  for(auto & parameter : parameter_robot_description->get_parameters({robot_description_ + "_planning/default_collision_operations"})){
```
I would like please some input about these changes @ahcorde

anasarrak committed:
Replace the spinner with SingpleThreadedExecutor

anasarrak committed:
Update getStateUpdateFrequency time, and add some TODO

anasarrak committed:
planning_scene, Time, params, publishers and canTransform(not finished)

anasarrak committed:
Update canTrasform passing TimePoint

anasarrak committed:
Change srv to shared, change planning_scene_world subscriber

anasarrak committed:
get scene (pointer)

anasarrak committed:
Update excludeRobotLinksFromOctree to use wallDuration

anasarrak committed:
Fix waitForCurrentRobotState duration

anasarrak committed:
Transform duration messages on initialize

anasarrak commented on moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp:
```diff
@@ -248,25 +248,27 @@ void PlanningSceneMonitor::initialize(const planning_scene::PlanningScenePtr& sc
 
   last_update_time_ = last_robot_motion_time_ = clock.now();
-  // last_robot_state_update_wall_time_.now() = std::chrono::system_clock::now();
+  last_robot_state_update_wall_time_ = clock.now();
```
@davetcoleman and @mlautman 

Im trying to change the ros::wallTime and ros::wallDuration, as was spoken with williams they doesn't seems to be ported to ros2. But the correct way to use it is by replacing ``std::chrono::system_clock::now()`` to ros::wallTime (or at least, this is what I underestood from https://github.com/ros-planning/moveit2/issues/31), also, I'm replacing ros::wallDuration by ``std::chrono::duration`` (such as std:chrono::seconds ...) .

Since there are a lot of time/walltime I would like to know if my procedure with that is the correct one, and if not, please help me with it.

mlautman replied:
This sounds reasonable to me. You should be able to create a quick prototype demonstrating that `ros::WallTime` acts equivalently to `std::chrono::system_clock`. Once you have proved that these are equivalent, or come up with a simple translation, it should be straightforward to do a find-replace across the code base. Note that you also need to update the includes accordingly. I would be more careful with ros::Time as it behaves differently if you are using simulated time vs real time and according to [this ROS Answers post](https://answers.ros.org/question/287946/ros-2-time-handling/) you need a node to be able to get simulated time which may require non-trivial modifications to the code

anasarrak replied:
After some testing, I'm going to be using the next code to replace the wallTime and to obtein the sec.

```
    auto time = std::chrono::system_clock::now();
    auto sec = std::chrono::duration<double>(time.time_since_epoch()).count();
``` 
Is it allright?

mlautman replied:
I would discourage the use of auto but otherwise it looks reasonable. I would encourage you to create tests before spending too much time on implementation.

anasarrak committed:
Fix last_robot_state_update_wall_time_ walltime

anasarrak committed:
wrong spelling get_topic_name

anasarrak committed:
get_topic_name

anasarrak committed:
Change dt_state_update and onStateUpdate function

anasarrak committed:
Changes on boost::algorithm, add TODOs, change time
Changed boost::algorithm to std::algorithm since c++11 supports it

Add TODO to collision_object_subscriber_, it usses message_filter,
so I have to look more into that since the compilation is crashing

Changes on stateUpdateTimerCallback to std::chrono::duration

anasarrak committed:
Minor fixes

anasarrak committed:
Change frequency duration on setStateUpdateFrequency()

anasarrak committed:
changes to ros_wall_timer

anasarrak committed:
changes on state_update_timer_
ros1 walltimer has implemented the start() and cancel() functions
http://docs.ros.org/indigo/api/roscpp/html/wall__timer_8cpp_source.html

for ros2 I tried to replace the start generating a new callback
and to replace the cancer I destroy the pointer. it's still a WIP
that has to be reviewed further

Looking into TimerBase class, there is a cancel() and reset()
functions, but they dont seem to be the same
https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/timer.cpp

anasarrak committed:
Add TODO to the namespace
Where does the namespace get declarated? Is it something related to the configs?

LanderU committed:
Update Dockerfile

Fix repos

LanderU committed:
Update Dockerfile

vcs fix

LanderU committed:
Fix CI on this branch

anasarrak committed:
Change robot_description parameters
is there any way of using hashmap with the parameters directly
as it's used in ros1?

anasarrak committed:
Merge branch 'planning_last' of github.com:AcutronicRobotics/moveit2 into planning_last

anasarrak committed:
rclcpp_error non trivial fix

anasarrak committed:
More non_trivial fix

anasarrak committed:
Update planning_scene_monitor demo_scene

anasarrak committed:
Starting with robot_model_loader

As it's needed for the demo_scene

anasarrak committed:
robot_model_loader, update parameters

anasarrak committed:
robot_model_loader, Update logger

anasarrak committed:
Change ros::time

anasarrak committed:
Fix the demo & update the Cmakelist for the demo

LanderU committed:
Change ws

anasarrak committed:
Update trajectory_monitor to compile for ros2

anasarrak committed:
Update namespace comment

anasarrak committed:
Fix robot_model_loader for non-trivial error

vmayoral committed:
moveit_ros/planning, updates for OS X

keeps failing, package needs further work

vmayoral commented:
@anasarrak, fixed a number of issues in OS X but the package is still failing:
```bash
/Users/victor/ros2_moveit_ws/src/moveit2/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp:938:54: error: cannot pass object of non-trivial type 'rclcpp::Duration' through variadic function; call will abort at runtime [-Wnon-pod-varargs]
    RCLCPP_DEBUG(logger, "last robot motion: %i ago",(t - last_robot_motion_time_));
                                                     ^
/Users/victor/ros2_moveit_ws/src/moveit2/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp:1184:79: warning: 'ConstPtr' is deprecated [-Wdeprecated-declarations]
void PlanningSceneMonitor::onStateUpdate( const sensor_msgs::msg::JointState::ConstPtr& /*joint_state */)
                                                                              ^
/Users/victor/ros2_ws/install/sensor_msgs/include/sensor_msgs/msg/joint_state__struct.hpp:138:11: note: 'ConstPtr' has been explicitly marked deprecated here
  typedef DEPRECATED_sensor_msgs_msg_JointState
          ^
/Users/victor/ros2_ws/install/sensor_msgs/include/sensor_msgs/msg/joint_state__struct.hpp:25:63: note: expanded from macro 'DEPRECATED_sensor_msgs_msg_JointState'
# define DEPRECATED_sensor_msgs_msg_JointState __attribute__((deprecated))
                                                              ^
/Users/victor/ros2_moveit_ws/src/moveit2/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp:1367:23: error: cannot pass object of non-trivial type 'std::__1::__vector_base<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >::value_type' (aka 'std::__1::basic_string<char>') through variadic function; call will abort at runtime [-Wnon-pod-varargs]
                      all_frame_names[i], target, ex.what());
                      ^
/Users/victor/ros2_moveit_ws/src/moveit2/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp:1367:43: error: cannot pass object of non-trivial type 'const std::string' (aka 'const basic_string<char, char_traits<char>, allocator<char> >') through variadic function; call will abort at runtime [-Wnon-pod-varargs]
                      all_frame_names[i], target, ex.what());
                                          ^
1 warning generated.
3 warnings and 3 errors generated.
make[2]: *** [planning_scene_monitor/CMakeFiles/moveit_planning_scene_monitor.dir/src/planning_scene_monitor.cpp.o] Error 1
make[1]: *** [planning_scene_monitor/CMakeFiles/moveit_planning_scene_monitor.dir/all] Error 2
make: *** [all] Error 2
```

are you currently looking at this errors?

anasarrak committed:
Fix planning_scene_monitor cmakelist

LanderU committed:
Update Dockerfile

* Added numpy-dev

anasarrak committed:
Update Logger

anasarrak committed:
Update clock and TODO

anasarrak committed:
Change dur time on haveCompleteState and TODO

Added a TODO for the ros2/geometry2 functions that are not ported yet
(_addTransformsChangedListener and _removeTransformsChangedListener)

Also modified the waitForCurrentState function to get the proper chrono
Duration

anasarrak committed:
Revert back const to waitForCurrentState

anasarrak committed:
Non-trivial error fix

anasarrak committed:
ros_planning export dependencies

anasarrak committed:
Update move_group cmakelist

anasarrak committed:
Update move_group package.xml and remove colcon_ignore

anasarrak committed:
Start with rdf_loader

anasarrak committed:
Update rdf_loader cmakelist

anasarrak committed:
Port rdf_loader to ROS2

anasarrak committed:
Port kinematics_plugin_loader to ROS2

anasarrak committed:
Update move_group cmakelists

anasarrak committed:
Update moveit_core controller_manager

anasarrak committed:
prepare plan_execution to compile

anasarrak committed:
update msg on sensor_manager

anasarrak committed:
update plan_execution logger and WallDuration

anasarrak committed:
Port move_group_capability to ros2

anasarrak committed:
Some changes to make planning_scene_interface_move_group compilable

Changes on:
 -planning_interface:
  CMakelists updated so it only compiles move_group_interface

 -move_group_interface:
Changes on cmakelist, msgs and commented out all the code, letting
available plan and execute, ill be focusing more on these changes once
planning_scene_monitor PR is closed.

 - planning_scene_interface:
Changes on the msgs on the .h, still has to be modified the .cpp

 - py_binding_tools:

 The cmakelist is prepared to compile with colcon.

anasarrak committed:
Update planning_scene_interface.cpp

vmayoral commented:
@anasarrak, what's the status of this?

anasarrak committed:
Re enable private variables and change getting clock duration

anasarrak committed:
Merge branch 'move_group' of github.com:AcutronicRobotics/moveit2 into move_group

anasarrak committed:
Update waitForAction function for ros2 and rename action variables

The waitForAction function has some callbacks that has to be reviewed,
added some TODO and a logger so it prints something when we launch it

anasarrak committed:
Rename geometry_msgs

anasarrak committed:
Change publishers and action calls

anasarrak committed:
Change wallduration to std::chrono on the header

anasarrak committed:
Cange service calls

anasarrak committed:
Add the correct declaration to the services

anasarrak committed:
Update query_service (getInterfaceDescription) callback

anasarrak committed:
Update callback

anasarrak committed:
Enable getRobotModel

anasarrak committed:
Change ros_fatal_stream_named to RCLCPP_FATAL

anasarrak committed:
Change trajectory_event_publisher_ and trajectory_event_publisher_

Changed both publishers to rclcpp publishers, Added a TODO, since
trajectory_execution_manager is not ported yet, I have hardcoded the
topic name

anasarrak committed:
change action node name

Commit:
Merge pull request #37 from AcutronicRobotics/move_group

Sync Move group with planning_last

anasarrak commented:
Due to the conflicts with master I'll split this pr on different PRs, since planning_scene_monitor needs to be ported for the different modules, for example plan_execution, trajectory_execution_manager... I prefer to merge it on master as it is and fix it on the future.

anasarrak committed:
Add TODO for the tests to collision_detection cmakelist

anasarrak committed:
Disable move_group_interface

anasarrak committed:
Remove ros1_bridge

anasarrak committed:
Update depends on package.xml (perception)
